File: java/src/strategy/implementations/Entropy.java
package strategy.implementations;

import java.util.ArrayList;
import java.util.List;

import strategy.Answer;
import strategy.Code;

/**
 * A strategy that seeks the guess with the maximum entropy.
 * 
 * In addition this strategy tries to minimize the travel distance.
 */
public class Entropy extends BasicStrategy {

	public Entropy(int codeLength) {
		super(codeLength);
	}

	@Override
	public Code guess(Answer answer) {
		this.removeInconsistentCodes(this.consistentCodes, this.lastGuess,
				answer);
		List<Code> bestGuesses = new ArrayList<Code>();
		bestGuesses.add(this.consistentCodes.get(0));
		double maxEntropy = 0;
		int totalSize = this.consistentCodes.size();
		for (Code code : this.allCodes) {
			double entropy = 0;
			for (Answer a : this.allAnswers) {
				int partSize = getConsistentCodes(this.consistentCodes, code, a)
						.size();
				if (partSize != 0) {
					double I = Math.log(1.0 * totalSize / partSize) / Math.log(2);
					double P = 1.0 * partSize / totalSize;
					entropy += entropy + I * P;
				}
			}
			if (entropy == maxEntropy && entropy > 0) {
				bestGuesses.add(code);
			}
			if (entropy > maxEntropy) {
				maxEntropy = entropy;
				bestGuesses.clear();
				bestGuesses.add(code);
			}
		}
		// Use, if possible, consistent codes
		List<Code> consistentBestGuesses = getConsistentCodes(bestGuesses,
				this.lastGuess, answer);
		if (!consistentBestGuesses.isEmpty()) {
			bestGuesses = consistentBestGuesses;
		}
		// Use a code with the shortest travel distance
		this.lastGuess = this.getShortestCode(bestGuesses);
		this.lastButton = this.lastGuess.get(this.CODE_LENGTH - 1);
		return this.lastGuess;
	}

	@Override
	public String toString() {
		return "Entropy";
	}

}

File: lua/strategies/entropy.lua
-- A strategy that seeks the guess with the maximum entropy.

local util = require 'strategies.util'
local createAllAnswers = util.createAllAnswers
local createAllCodes = util.createAllCodes
local getConsistentCodes = util.getConsistentCodes
local getShortestCode = util.getShortestCode
local min = math.min
local log = math.log

-- Main logic

local allCodes = createAllCodes(CODE_LENGTH)
local allAnswers = createAllAnswers(CODE_LENGTH)
local consistentCodes = {}
local lastGuess = nil
local lastButton = {x=0,y=0}

local function reset()
  for i,code in ipairs(allCodes) do
    consistentCodes[i] = code
  end
  lastGuess = getShortestCode(lastButton, consistentCodes)
  lastButton = lastGuess[CODE_LENGTH]
  return lastGuess
end

local function guess(blacks, whites)
  consistentCodes = getConsistentCodes(consistentCodes, lastGuess, 
      blacks, whites)
  local bestGuesses = {consistentCodes[1]}
  local totalSize = #consistentCodes
  local maxEntropy = 0
  for _,code in ipairs(allCodes) do
    local entropy = 0
    for _,answer in ipairs(allAnswers) do
      local partSize = #getConsistentCodes(consistentCodes, code, 
          answer.blacks, answer.whites)
      if partSize ~= 0 then
        local I = log(totalSize / partSize) / log(2)
        local P = partSize / totalSize
        entropy = entropy + I * P
      end
    end
    if entropy == maxEntropy and entropy > 0 then
      bestGuesses[#bestGuesses+1] = code 
    end
    if entropy > maxEntropy then
      maxEntropy = entropy
      bestGuesses = {code}
    end
  end
  -- Use, if possible, consistent codes
  local consistentBestGuesses = getConsistentCodes(bestGuesses, lastGuess, 
      blacks, whites)
  if #consistentBestGuesses ~= 0 then bestGuesses = consistentBestGuesses end
  -- Use a code with the shortest travel distance
  lastGuess = getShortestCode(lastButton, bestGuesses)
  lastButton = lastGuess[CODE_LENGTH]
  return lastGuess
end

return {
  reset = reset,
  guess = guess
}
